rfd/0064-audit-modes.md

```diff
@@ -0,0 +1,190 @@
+---
+authors: Gabriel Corado (gabriel.oliveira@goteleport.com)
+state: draft
+---
+
+# RFD 64 - Audit modes
+
+## What
+
+This RFD proposes giving users a configuration to decide how to deal with
+sessions that can't be recorded or have audit logging failures.
+
+## Why
+
+In cases where Teleport is configured to record sessions on the node using the
+asynchronous format, disk failures (for example, no space left) causes the node
+to be inaccessible. Where Teleport is the only way of accessing those resources,
+administrators cannot perform the necessary actions on these nodes to bring them
+back to a normal state.
+
+## Goals
+
+* Prevent nodes from being inaccessible when there are audit errors.
+* Enable users to configure the strictness of audit failures.
+
+**Non-goals:**
+
+* Infrastructure-related metrics (such as disk space left).
+* Enable access when the recording mode is set to the proxy.
+
+## Details
+
+Audit modes will define how Teleport proceeds in case of audit failures. Those
+failures vary depending on how session recording is configured:
+
+| Recording mode                 | Failures handled                                             |
+| ------------------------------ | ------------------------------------------------------------ |
+| `node` and `proxy`             | I/O errors while writing audit logs/session recording data on the server disk. |
+| `node-async` and `proxy-async` | Connection errors with Proxy/Auth server while streaming audit logs/session recording data. |
+
+There are going to be two audit modes:
```

Comment:
> Audit events are always emitted directly to the auth server and never written to disk first, right?

Not really. In some cases, they're also written into the disk. We use a [`TeeStreamer`](https://github.com/gravitational/teleport/blob/master/lib/events/emitter.go#L457) when the recording mode is not sync (like [in the SSH sessions](https://github.com/gravitational/teleport/blob/master/lib/srv/sess.go#L1260-L1263)), so those events are sent to the disk and auth server ([except for a few events](https://github.com/gravitational/teleport/blob/master/lib/events/emitter.go#L501)). For example, looking at the session recording we can see the [start](https://github.com/gravitational/teleport/blob/master/lib/srv/sess.go#L999) and [end](https://github.com/gravitational/teleport/blob/master/lib/srv/sess.go#L1216) events:

```
$ tsh play -f json <session-id>
{"ei":0,"event":"session.start", ...}
{"ei":1,"event":"print", ...}
...
{"ei":10,"event":"print", ...}
{"ei":11,"event":"session.end", ...}
```

However, failing to emit an audit event will only log an error message since they're already sent to auth server.

> If this is true, then running out of disk isn't so much of an issue for audit events.

Right, this is not an audit events issue. But the RFD mentions it will also cover errors while emitting events to auth server (we can consider removing this part if it makes sense).